synapseclient/table.py

```diff
@@ -440,21 +440,27 @@ def _csv_to_pandas_df(
 
     line_terminator = str(os.linesep)
 
+    pandas_args = {
+        "dtype": dtype,
+        "sep": separator,
+        "quotechar": quote_char,
+        "escapechar": escape_char,
+        "header": 0 if contain_headers else None,
+        "skiprows": lines_to_skip,
+        **kwargs,
+    }
+    pandas_args.update(kwargs)
```

Comment:
What happens when you add a kwargs pandas doesn't accept? Does to throw a pandas error?

Not saying we need to catch it, just curious

Reply:
Yeah it would throw a `pandas` error like

```
Traceback (most recent call last):
  File "/Users/bmacdonald/Desktop/test_duplicate_kwargs.py", line 28, in <module>
    test_table_df_sep = test_table.asDataFrame(test=",")
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bmacdonald/Development/repos/synapsePythonClient/synapseclient/table.py", line 2526, in asDataFrame
    return _csv_to_pandas_df(
           ^^^^^^^^^^^^^^^^^^
  File "/Users/bmacdonald/Development/repos/synapsePythonClient/synapseclient/table.py", line 458, in _csv_to_pandas_df
    df = pd.read_csv(
         ^^^^^^^^^^^^
TypeError: read_csv() got an unexpected keyword argument 'test'
```

which I chose not to handle because if the arguments `read_csv` change we would have to update how we handle it

Reply:
`separator` is a default `_csv_to_pandas_df` argument so if you pass that to `asDataFrame` it will error with 
```
Traceback (most recent call last):
  File "/Users/bmacdonald/Desktop/test_duplicate_kwargs.py", line 28, in <module>
    test_table_df_sep = test_table.asDataFrame(separator=",", sep=",")
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bmacdonald/Development/repos/synapsePythonClient/synapseclient/table.py", line 2526, in asDataFrame
    return _csv_to_pandas_df(
           ^^^^^^^^^^^^^^^^^^
TypeError: synapseclient.table._csv_to_pandas_df() got multiple values for keyword argument 'separator'
```

Whereas `sep` would overwrite the default `separator` argument because it is a `pandas.read_csv` argument rather than a `_csv_to_pandas_df` arg.

Reply:
Right, I guess my question is more of a "philosophical" one. Do we accept the `sep="\t"` or do we stick with `separator=","` because that is the defined argument and not `kwargs`?

That said, when this is refactored, I envision there to not be a `seperator` parameter at all.

Reply:
I was approaching this with a bias towards intentionality. If a user chooses to set `sep="\t"`, that should override the default `separator` argument set in `_csv_to_pandas_df` because that is probably what a user would expect to happen. If we enforce the default behavior we would need to provide some message to say that this is happening or there would be a confusing silent failure happening from the user's perspective.

I also imagine that once this is refactored this decision will be abstracted away, anyway.

Reply:
Right, what are the changes a user would set

`separator` to be something different? Should we add a blurb saying that the kwargs take precedence?

Reply:
> Right, what are the chances a user would set `separator` to be something different? Should we add a blurb saying that the kwargs take precedence? Like
> 
> Separator='L' Sep='M'
> 
> It would take sep

+1, perhaps we add to each function argument what `kwargs` would replace it. See my example for `separator` for my initial thought:

```
    """
    Convert a csv file to a pandas dataframe
    Arguments:
        filepath: The path to the file.
        separator: The separator for the file, Defaults to `DEFAULT_SEPARATOR`. Passed as `sep` to pandas. If `sep` is supplied as a `kwarg` it will be used instead of this `separator` argument.
        quote_char: The quote character for the file,
          Defaults to `DEFAULT_QUOTE_CHARACTER`.
        escape_char: The escape character for the file,
                    Defaults to `DEFAULT_ESCAPSE_CHAR`.
                    contain_headers: Whether the file contains headers,
                    Defaults to `True`.
        lines_to_skip: The number of lines to skip at the beginning of the file,
                        Defaults to `0`.
        date_columns: The names of the date columns in the file
        list_columns: The names of the list columns in the file
        rowIdAndVersionInIndex: Whether the file contains rowId and
                                version in the index, Defaults to `True`.
        dtype: The data type for the file, Defaults to `None`.
        **kwargs: Additional keyword arguments to pass to pandas.read_csv. See
                    https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html
                    for complete list of supported arguments.
    Returns:
        A pandas dataframe
    """
```

Reply:
Well users shouldn't be calling `_csv_to_pandas_df` directly, it's called by `asDataFrame` so if they set `separator` and `sep` as `kwargs` in `asDataFrame` you would get this error [above](https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1132#discussion_r1771956172). You shouldn't really encounter a situation where someone is doing that though because that would be a pretty weird thing to do. We can add something saying that the `kwargs` take precendence but it would be kind of confusing potentially because `_csv_to_pandas_df` and its default arguments are somewhat hidden.

Reply:
We could do what @BryanFauble suggests, but then we are maintaining documentation on `pandas` arguments which could potentially change in a future version which I don't love. But happy to do it if we think it's necessary.

Reply:
@BWMac Should we add these types of comments to the top level where a user enters into this logic. That way everything is explicitly stated for how the function arguments are going to be evaluated.

Some areas at a glance:

1. https://github.com/Sage-Bionetworks/synapsePythonClient/blob/857154e45ddd9f85425b7417d349786da633e8f2/synapseclient/table.py#L2414-L2427
2. https://github.com/Sage-Bionetworks/synapsePythonClient/blob/857154e45ddd9f85425b7417d349786da633e8f2/synapseclient/table.py#L2476-L2496

Reply:
> so if they set separator and sep as kwargs in asDataFrame

This might be possible because one is set in the class instance constructor, and the other is passed in via the function argument.

Reply:
Ah right that's a really good point. I like the idea of putting this information closer to the top level so users will see it. I will take a look later and make some updates to documentation to help users navigate this. Thanks for the input @BryanFauble and @thomasyu888 !

Reply:
Updated docstrings in latest commit